staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go

```diff
@@ -45,26 +51,31 @@ var (
 // adjustClusterNameIfWildcard determines the logical cluster name. If this is not a wildcard list/watch request,
 // the cluster name is returned unmodified. Otherwise, the cluster name is extracted from the key based on whether
 // it is a CR partial metadata request (prefix/identity/clusterName/remainder) or not (prefix/clusterName/remainder).
-func adjustClusterNameIfWildcard(cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
+func adjustClusterNameIfWildcard(shard genericapirequest.Shard, cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
 	if cluster.Name != logicalcluster.Wildcard {
 		return cluster.Name
 	}
 
 	keyWithoutPrefix := strings.TrimPrefix(key, keyPrefix)
 
 	var regex *regexp.Regexp
-
-	if cluster.PartialMetadataRequest && crdRequest {
-		regex = crdWildcardPartialMetadataClusterNameRegex
-	} else {
-		regex = wildcardClusterNameRegex
+	if !shard.Wildcard() {
+		if cluster.PartialMetadataRequest && crdRequest {
+			regex = crdWildcardPartialMetadataClusterNameRegex
+		} else {
+			regex = wildcardClusterNameRegex
+		}
+		matches := regex.FindStringSubmatch(keyWithoutPrefix)
+		if len(matches) >= 2 {
+			return logicalcluster.New(matches[1])
+		}
+		return logicalcluster.Name{}
 	}
 
-	matches := regex.FindStringSubmatch(keyWithoutPrefix)
-	if len(matches) >= 2 {
-		return logicalcluster.New(matches[1])
+	matches := wildcardShardClusterNameRegex.FindStringSubmatch(keyWithoutPrefix)
```

Comment:
Am still confused here. Why is this not `wildcardClusterNameRegex`?

Reply:
we cannot use `wildcardClusterNameRegex` because it won't capture a cluster name, only a shard name.
for example: given the key of `amber/root:org:ws/some-name` it will capture `amber` as a cluster name

basically, once this PR merges we will have two ways of storing data.
1. the kcp server that always adds a cluster name 
2. the cache-server that always adds a shard name and a cluster name.

when reading data back we need to take those two modes into account.

Reply:
that means the cache-server won't read data stored by the kcp server and vice versa.

Reply:
the above only applies to shard = "*", when a shard is specified then we change the prefix (https://github.com/kcp-dev/kubernetes/pull/96/files#diff-e52c66f0ffc19a41478dc44cd39fc05fe64f371d73f9db362c9df6047b94b810R239) and if the cluster was not specified will use `wildcardClusterNameRegex`

Reply:
and finally if a shard and a cluster were specified we will simply
```
f cluster.Name != logicalcluster.Wildcard {
		return cluster.Name
}
```

I want to capture all those details in https://github.com/kcp-dev/kcp/issues/1866

Reply:
ack. So the old case is in the if above.